### ISSUE TITLE
Add request batching so a model can serve multiple queries in one forward pass

### ISSUE THREAD

**Issue by nicolaslara** [tracel-ai/burn-lm#56]
While digging through the inference path I realized that right now a model can only serve **one request at a time**. The registry wires every model with `MutexChannel`, and `run_job` holds the lock for the *entire* generation:

```rust
fn run_job(&self, job: InferenceJob) -> InferenceResult<Stats> {
    let mut server = self.server.lock().unwrap(); // locked for the whole generation
    server.run_job(job)
}
```

So when two HTTP requests come in, the second one just sits behind the lock. The GPU is effectively running at **batch size 1** the whole time (small matmuls, weights read from HBM for a single sequence) which is nowhere near what the hardware can do.

Fixing this would be a pre-requisite for adding vLLM style *continuous batching*.

I'd love to work on this if it's something that's wanted / will get merged.

## Suggested Design

Add batching as an **opt-in path** that leaves simple models (Parrot) and all front-ends untouched.

**Stays the same**

- The job/emitter/listener protocol.
- `InferencePlugin::run_job(&self, job)` keeps its signature. The HTTP server already
  drives N concurrent requests on N blocking threads → N simultaneously-pending jobs, so
  no front-end changes needed.

**Change 1: a `BatchingChannel` that owns a scheduler.** Instead of locking per call, a long-lived worker thread owns the server `&mut` and runs a continuous-batching loop. Callers submit into a queue and block only on their own job finishing (via a per-job completion signal).

**Change 2: split the model into a server + reusable decoder primitive.** The server keeps doing lifecycle/config/prompt handling, but the actual generation path moves through a decoder primitive that can be driven by either a single-request engine or a continuous batching engine.

```rust
pub trait BatchedDecoder {
    type Cache;

    fn allocate_cache(&self, capacity: BatchCapacity) -> Self::Cache;

    fn forward(
        &mut self,
        batch: ForwardBatch,
        cache: &mut Self::Cache,
    ) -> ForwardOutput; // logits + cache/state updates
}

pub trait BatchedInferenceServer: InferenceServer {
    type Decoder: BatchedDecoder;

    fn decoder(&mut self) -> InferenceResult<&mut Self::Decoder>;
}
```

The important part is that model authors shouldn't have to write two separate generation loops. The single-job path can use the same decoder via a `SingleRequestEngine`, while the batching path uses a `ContinuousBatchEngine` owned by the `BatchingChannel` worker:

`prefill` and `decode` are engine concepts, not required model-trait methods. The engines call `decoder.forward(...)` with different shapes: a full prompt for prefill (`[batch, prompt_seq]`) and the latest token for decode (`[batch, 1]`). If we later need specialized prefill/decode paths for a model, they can be added as optional hooks with default implementations over `forward`.

```rust
pub struct Llama3InstructServer {
    config: Llama3ServerConfig,
    decoder: Option<LlamaDecoder<Tiktoken>>,
}

impl InferenceServer for Llama3InstructServer {
    fn load(&mut self) -> InferenceResult<Option<Stats>> {
        if self.decoder.is_none() {
            self.decoder = Some(load_llama_decoder(...)?);
        }
        Ok(stats)
    }

    fn run_job(&mut self, job: InferenceJob) -> InferenceResult<Stats> {
        self.load()?;
        let prompt = self.prompt_from_job(job.task)?;
        let decoder = self.decoder.as_mut().unwrap();

        SingleRequestEngine::new(decoder, self.config.generation())
            .run(prompt, job.emitter)
    }
}

impl BatchedInferenceServer for Llama3InstructServer {
    type Decoder = LlamaDecoder<Tiktoken>;

    fn decoder(&mut self) -> InferenceResult<&mut Self::Decoder> {
        self.load()?;
        Ok(self.decoder.as_mut().unwrap())
    }
}
```

So the flow is:

```text
single request:
  InferenceServer::run_job
    -> SingleRequestEngine<Decoder>

batched serving:
  BatchingChannel worker
    -> ContinuousBatchEngine<Decoder>
```

Same decoder, two engines. The registry picks `BatchingChannel` when a model implements `BatchedInferenceServer`, otherwise falls back to `MutexChannel`.

**Prerequisite — use Burn's attention op in Llama.** Before adding batching, we should update Llama attention to use Burn's backend-dispatched attention tensor op instead of reimplementing attention internally with `matmul -> softmax -> matmul`. Burn now has attention implemented across backends, and batching should build on top of that lower-level primitive rather than maintaining a separate attention implementation in Burn-LM.

## Rough plan 

- [ ] Explore the implementation shape with a small prototype: split Llama into a
      server + `LlamaDecoder`, and validate that `SingleRequestEngine<Decoder>` can
      reproduce today's single-request behavior.
- [x] Update Llama attention to use Burn's backend-dispatched attention op.
- [ ] Add `BatchedDecoder`, `BatchCapacity`, `Batch`, and per-job completion signal in
      `burn-lm-inference` (no behavior change yet).
- [x] Add `BatchedInferenceServer` + `BatchingChannel` with a `ContinuousBatchEngine`
      worker; gate model selection on the trait.
- [ ] Make Llama's generation loop batch-aware: ragged batch, per-sequence positions +
      stop detection, prefill/decode interleaving. Validate it reproduces current
      single-sequence outputs.
- [ ] Continuous-batching admission on a KV/token budget; benchmark tokens/sec under
      concurrent load vs the `MutexChannel` baseline.
- [ ] Async submit API so we're not parking one OS thread per in-flight request
      at high concurrency.
- [ ] finish `SseGenerationListener` so batched per-token output streams as real
      OpenAI SSE chunks.

## Open questions

- Scheduling policy: FCFS vs priority vs max-wait (latency/throughput trade-off).
- Determinism: batched runs need to match single-sequence numerics so existing tests stay meaningful.
- Cancellation: a dropped connection should evict its sequence from the batch.

## Scope

I'd like to take this on. I'd start by exploring the implementation in Llama first: split out a decoder primitive, make sure the single-request path can be expressed through a `SingleRequestEngine`, then move the trait/channel scaffolding into `burn-lm-inference` once that shape feels right.
